osler/inventory/test_views.py

```diff
@@ -0,0 +1,127 @@
+                            'category': DrugCategory.objects.last().pk,
+                            'manufacturer': Manufacturer.objects.last().pk
+                        }
+
+    def test_drug_list_view(self):
+        drug = Drug.objects.create(**drug_dict())
+        url = reverse('inventory:drug-list')
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+        self.assertIn(drug, response.context_data['object_list'])
+        self.assertTemplateUsed(response, 'inventory/inventory-main.html')
+
+    def test_drug_update(self):
+        drug = Drug.objects.create(**drug_dict())
+        n_drugs = len(Drug.objects.all())
+        url = reverse('inventory:drug-update', kwargs={'pk':drug.pk})
+        response = self.client.post(url, self.diff_drug, follow=True)
+        drug_get = Drug.objects.get(pk=drug.pk)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(n_drugs, Drug.objects.count())
+        for param in self.diff_drug:
+            self.assertEqual(str(self.diff_drug[param]),
+                             str(getattr(drug_get, param)))
+
+    def test_drug_dispense_can_dispense(self):
+        drug = Drug.objects.create(**drug_dict())
+        drug_initial = Drug.objects.get(pk=drug.pk)
+        url = reverse('inventory:drug-dispense')
+        response = self.client.post(url, {'pk':drug.pk,'num':'5'}, follow=True)
+        drug_final = Drug.objects.get(pk=drug.pk)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(str(getattr(drug_initial,'stock')-5),
+                         str(getattr(drug_final,'stock')))
+
+    def test_drug_dispense_cannot_dispense(self):
+        drug = Drug.objects.create(**drug_dict())
+        drug_initial = Drug.objects.get(pk=drug.pk)
+        url = reverse('inventory:drug-dispense')
+        response = self.client.post(url, {'pk':drug.pk,'num':'11'}, follow=True)
+        drug_final = Drug.objects.get(pk=drug.pk)
+        self.assertEqual(response.status_code, 404)
+
+class TestDrugAdd(TestCase):
+
+    fixtures = ['core', 'workup', 'inventory']
+
+    def setUp(self):
+        log_in_provider(self.client, build_provider())
+
+        self.drug_dict = {
+                            'name': 'Somedrug',
+                            'unit': MeasuringUnit.objects.first(),
+                            'dose': 1000.0,
+                            'stock': 10,
+                            'expiration_date': '2040-01-01',
+                            'lot_number': 'ABCDEFGH',
+                            'category': DrugCategory.objects.first(),
+                            'manufacturer': Manufacturer.objects.first()
+                        }
+
+    def test_pre_drug_add_new_with_collision(self):
+        drug = Drug.objects.create(**self.drug_dict)
+        url = reverse('inventory:pre-drug-add-new')
+        self.drug_dict['manufacturer'] = Manufacturer.objects.first().pk
```

Comment:
Hmmm, yes this is strange. I tried taking out the .pk and I got this error message:
`FAILED osler/inventory/test_views.py::TestDrugAdd::test_pre_drug_add_new_no_collision - AssertionError: 'AAH' != <Manufacturer: AAH>`
What's going on is that first a drug is created with `drug = Drug.objects.create(**self.drug_dict)`. After, I am trying to create another drug with the same name, lot number, and manufacturer, and if I don't add the .pk, it gives me the error above. If you have any suggestions on how I can improve this, let me know! At least the good news is that the test is doing what it's meant to do, to catch drugs with collisions. I am very open to suggestions since I've struggled with this while writing the test and this was the only solution I could come up with.